sklearn/linear_model/ridge.py

```diff
@@ -1087,56 +1400,42 @@ def fit(self, X, y, sample_weight=None):
             X, y, self.fit_intercept, self.normalize, self.copy_X,
             sample_weight=sample_weight)
 
-        gcv_mode = self.gcv_mode
-        with_sw = len(np.shape(sample_weight))
-
-        if gcv_mode is None or gcv_mode == 'auto':
-            if sparse.issparse(X) or n_features > n_samples or with_sw:
-                gcv_mode = 'eigen'
-            else:
-                gcv_mode = 'svd'
-        elif gcv_mode == "svd" and with_sw:
-            # FIXME non-uniform sample weights not yet supported
-            warnings.warn("non-uniform sample weights unsupported for svd, "
-                          "forcing usage of eigen")
-            gcv_mode = 'eigen'
+        gcv_mode = _check_gcv_mode(X, self.gcv_mode)
 
         if gcv_mode == 'eigen':
-            _pre_compute = self._pre_compute
-            _errors = self._errors
-            _values = self._values
+            decompose = self._decompose_gram
```

Comment:
Naming `self._decompose_gram` and `self._decompose_covariance_*` is disconnected from `gcv_mode`, which controls which path to take. Maybe `self._decompose_eigen` and `self._decompose_svd_*` (and the same for `self._solve_covariance`.

Reply:
I would rather preserve the name of the matrix that is formed in the method name. Maybe we could be even more explicit and use:

```
self._svd_decompose_covariance
```
and:

```
self._eigen_decompose_gram
```

to know the size of the intermediate datastructures depending on `n_samples` and `n_features`.

Reply:
I have made the change you suggest, and also changed the name of the method
which computes an svd of X by replacing "covariance" with "design_matrix".
so the names are now

- `_eigen_decompose_gram` (gcv_mode='eigen', eigen decomposition of gram matrix)
- `_svd_decompose_covariance` (gcv_mode='svd', eigen decomposition of covariance matrix, used when X is sparse)
- `_svd_decompose_design_matrix` (gcv_mode='svd', svd decomposition of design matrix, used when X is dense)

it is still not ideal because the second one is called `_svd...` and actually
computes an eigendecomposition, but this comes from the parameter choices used
to determine wether we solve the primal or dual problem, "svd" and "eigen"

Reply:
I had not realized the naming were so misleading.

Maybe we should actually rename those `gcv_mode` options to "gram and "covariance" to make it explicit which (while keeping backward compat with the old names with a deprecation warning).

BTW do you know if using a real svd on the design matrix would yield good results? Probably not for this PR but I am curious.

Reply:
an SVD of the design matrix is indeed what we use when X is dense and n > p .
When X is sparse and n > p we don't want to actually store the left singular
vectors of X because they are a dense matrix of the same shape as X.
So we do something very similar but instead of writing the hat matrix as
(ignoring the intercept and sample weights for simplicity):

X = U S V^T
H = X(X^TX + alpha I)X^T = U (S^2 / (S^2 + alpha I)) U^T

we write it as

H = XV (1 / (S^2 + alpha I)) V^TX^T

and since we only need the dot product with y and the diagonal of this matrix 
we avoid creating a dense matrix of size (n x p).
(the diagonal is computed by batch in `_sparse_multidot_diag`)
So it is very similar to using an SVD of X, but in practice we only compute an
eigendecomposition of X^TX. We can consider that we are implicitly using an SVD
of X, writing U = XVS^-1 instead of explicitly forming U. This is being added
in this PR, as the sparse low-dimensional case was not handled before, so the
naming problem of choosing "svd" and computing an eigendecomposition does not
arise on master. (by the way, if you see a better way to compute Hy and diag(H)
when X is sparse I can change this implementation).

as you said, the name of the matrix that is being decomposed is more informative
than "svd" or "eigen", which can be deduced from whether the matrix is square or
not.

finally, I don't know why one would use another setting than "auto", besides
testing and checking that results are similar.

also note that all methods will result in memory problems if X is sparse and
large in both dimensions (n and p); in this case it is probably better not to
use generalized cross-validation or any decomposition of X.